torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -211,39 +211,43 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
       [autogradContext, outputEdges, accumulateGradFuture](
           const variable_list& grads,
           const c10::optional<torch::utils::FutureError>& error) {
-        if (error) {
-          // Don't accumulate gradients if we receive an error.
-          // We must add the node information here since DistEngine::execute
-          // waits on accumulateGradFuture and will throw an exception once we
-          // set the error below.
-          std::string errorMsg = c10::str(
-              "Error on Node ",
-              DistAutogradContainer::getInstance().getWorkerId(),
-              ": ",
-              error->what());
-          accumulateGradFuture->setError(errorMsg);
-          return;
-        }
+        try {
+          if (error) {
```

Comment:
Catching exceptions is reasonable, but can we maybe be slightly more selective?

As an example, it doesn't seem there's a real reason for try/catch in lines 214...229 (similar in some other callback blocks).

Yes, it's true that setError() can throw if it has a "naughty" callback that throws. But then the solution here is to just setErrorIfNeeded again [which will do nothing since the callbacks have all been depleted by the original setError call]. This ends up not solving the problem, and maybe making things slightly worse - since we now started hiding an error, rather than causing it to bubble to a higher layer (where there's a greater chance we might notice it).

But, yes, it sounds great to try/catch around accumulateGrad, circa line 242, below. Perhaps even with a LOG(INFO) too if it errors out.